src/mavsdk/core/system_impl.h

```diff
@@ -25,6 +25,12 @@
 #include <mutex>
 #include <future>
 
+template<class... Ts> struct overloaded : Ts... {
+    using Ts::operator()...;
+};
+
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
```

Comment:
So this helper can only be used library internally. Was that the aim? Or should it be usable outside of the library in the public API?

And what was your use case for it?

Reply:
Still learning about std::variant, I found std::visit which had an example of using overloaded functions to handle the different active types of an std::variant: https://en.cppreference.com/w/cpp/utility/variant/visit

It's supposed to be used in plugin implementations, but could also be used outside if the user wants to. I'll send you an example very soon.

However, this approach has a fundamental flaw, that you could write a handler for uint32_t which would also accept types like uint8_t or uint16_t instead of dismissing it as wrong parameter type. Need to do some more testing about this.

Reply:
An example looks like this. If the plugin doesn't know the type necessary to run the callback function, because it's supplied by the user, this approach is pretty helpful:

Plugin code:

```
    template <class T>
    void add_param(const std::string& name, T default_value, const typename std::function<void(T t)>& callback) {
        _parent->provide_server_param_int(name, default_value);
        _parent->subscribe_param(
            name,
            overloaded {  
                [&name](auto value) { std::cerr << "invalid parameter type for " << name << std::endl; },
                callback
            },
            nullptr);
    }
```

User code:

```
camera_manager.add_param<uint32_t>("CAM_ISO", 0, [](uint32_t iso) { std::cout << "NEW ISO: " << iso << std::endl; });
```

Note that `_parent->provide_server_param_int` would also have to be replaced with a `_parent->provide_server_param`.

Reply:
I actually hate the fact that we have all types smaller than 32 bits. We could have saved quite a bit of trouble by not suggesting them in the spec. That was just stupid, I think.

Anyway, I see. In that case you need to add the `overloaded` helper template in one of the public headers, maybe in the param plugin.h.

Reply:
> Anyway, I see. In that case you need to add the overloaded helper template in one of the public headers, maybe in the param plugin.h.

Done!